File: lua/redirect/translate.lua
-- supported redirect status codes
local codes = {
    ["301"] = 301,
    ["302"] = 302,
    ["permanently"] = 301,
    ["temporarily"] = 302,
    default = 302
}

-- escape pattern characters
local function escape (s)
    return string.gsub(s, "[%^%$%(%)%%%.%[%]%*%+%-%?]", "%%%0")
end

local function translate (uri, args)
    if not uri then return end
    if not args then return end

    local target = args.target
    if not target then return end

    local status = codes[args.status] or codes.default
    -- no `path` assumes catch-all, so redirect immediately to `target`
    if not args.path then return target, status end

    local path, count = string.gsub(escape(args.path), "%%%*", "(.*)", 1)
    path = "^" .. path .. "$"

    -- if we can't find the pattern, break to continue to next record
    if not string.find(uri, path) then return end

    -- wildcard replacement of `uri` if there's a wildcard in our `path`
    if count then
        target = string.gsub(target, "%*", "%%1", 1)
        target = string.gsub(uri, path, target, 1)
    end

    return target, status
end

return translate

File: translate.go
package main

import (
	"bytes"
	"regexp"
	"strings"
)

type Redirect struct {
	Location string
	Status   int
}

func Translate(uri string, config *Config) *Redirect {
	if uri == "" {
		return nil
	}
	if config == nil {
		return nil
	}
	if config.To == "" {
		return nil
	}

	redirect := &Redirect{Location: config.To}

	switch config.RedirectState {
	case "301", "permanently":
		redirect.Status = 301
	case "302", "temporarily":
		redirect.Status = 302
	default:
		redirect.Status = 302
	}

	// no `From` assumes catch-all, so redirect immediately to `Location`
	if config.From == "" {
		return redirect
	}

	count := strings.Count(config.From, `*`)

	var exp bytes.Buffer
	exp.WriteString(`^`)
	exp.WriteString(strings.Replace(regexp.QuoteMeta(config.From), `\*`, `(.*)`, 1))
	exp.WriteString(`$`)

	fromRE := regexp.MustCompile(exp.String())

	// if we can't find the pattern, return to continue to next record
	if !fromRE.MatchString(uri) {
		return nil
	}

	// wildcard replacement of `uri` if there's a wildcard in our `From` path
	if count > 0 {
		redirect.Location = fromRE.ReplaceAllString(uri, strings.Replace(redirect.Location, `*`, `${1}`, 1))
	}

	return redirect
}
